api/auth_middleware.go

```diff
@@ -24,57 +34,127 @@ func getServiceAccountID(ctx context.Context) (string, bool) {
 }
 
 // authMiddleware authenticates either access_token or key pair
-func authMiddleware(
-	sasUC usecases.ServiceAccounts,
-) func(http.Handler) http.Handler {
+func authMiddleware(sasUC usecases.ServiceAccounts) func(http.Handler) http.Handler {
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			authorization := r.Header.Get("authorization")
-			parts := strings.Split(authorization, " ")
-			if authorization == "" || len(parts) != 2 {
+			header := r.Header.Get("authorization")
+			authHeaderContents := strings.Split(header, " ")
+			authHeader := buildAuth(header, authHeaderContents[0], authHeaderContents[1])
+			if !authHeader.isValid() {
 				w.WriteHeader(http.StatusUnauthorized)
 				return
 			}
+
 			var ctx context.Context
-			l := middleware.GetLogger(r.Context())
-			if parts[0] == "KeyPair" {
-				keyPair := strings.Split(parts[1], ":")
-				accessKeyPairAuth, err := sasUC.WithContext(r.Context()).
-					AuthenticateKeyPair(keyPair[0], keyPair[1])
-				if err != nil {
-					l.WithError(err).Error("auth failed")
-					w.WriteHeader(http.StatusInternalServerError)
-					return
-				}
-				w.Header().Set("x-service-account-name", accessKeyPairAuth.Name)
-				ctx = context.WithValue(r.Context(), serviceAccountIDCtxKey, accessKeyPairAuth.ServiceAccountID)
-			} else if parts[0] == "Bearer" {
-				accessToken := parts[1]
-				accessTokenAuth, err := sasUC.WithContext(r.Context()).
-					AuthenticateAccessToken(accessToken)
-				if err != nil {
-					l.WithError(err).Info("auth failed")
-					if _, ok := err.(*errors.EntityNotFoundError); ok {
-						w.WriteHeader(http.StatusUnauthorized)
-						return
-					}
-					l.Error(err)
-					w.WriteHeader(http.StatusInternalServerError)
-					return
-				}
-				w.Header().Set("x-email", accessTokenAuth.Email)
-				if accessTokenAuth.AccessToken != accessToken {
-					w.Header().Set("x-access-token", accessTokenAuth.AccessToken)
-				}
-				ctx = context.WithValue(
-					r.Context(), serviceAccountIDCtxKey, accessTokenAuth.ServiceAccountID,
-				)
+			var err error
+			logger := middleware.GetLogger(r.Context())
+
+			if authHeader.isKeyPair() {
```

Comment:
You can use switch with: authHeader.Type
switch authHeader.Type {
case models.AuthenticationTypes.KeyPair:
case models.AuthenticationTypes.OAuth2:

Can you test?

Reply:
Good catch! Made the change in 15d0d9cad8dd13bb4f29b1c1dc20bb65794ce9d3